src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardRca.java

```diff
@@ -106,14 +110,18 @@ private void consumeFlowUnit(final MetricFlowUnit metricFlowUnit, final String m
                                  final HashMap<IndexShardKey, SlidingWindow<SlidingWindowData>> metricMap) {
         for (Record record : metricFlowUnit.getData()) {
             try {
-                IndexShardKey indexShardKey = IndexShardKey.buildIndexShardKey(record);
-                double usage = record.getValue(MetricsDB.SUM, Double.class);
-                SlidingWindow<SlidingWindowData> usageDeque = metricMap.get(indexShardKey);
-                if (null == usageDeque) {
-                    usageDeque = new SlidingWindow<>(SLIDING_WINDOW_IN_SECONDS, TimeUnit.SECONDS);
-                    metricMap.put(indexShardKey, usageDeque);
+                String indexName = record.getValue(INDEX_NAME.toString(), String.class);
+                Integer shardId = record.getValue(SHARD_ID.toString(), Integer.class);
+                if (indexName != null &&  shardId != null) {
+                    IndexShardKey indexShardKey = IndexShardKey.buildIndexShardKey(record);
+                    double usage = record.getValue(MetricsDB.SUM, Double.class);
+                    SlidingWindow<SlidingWindowData> usageDeque = metricMap.get(indexShardKey);
+                    if (null == usageDeque) {
+                        usageDeque = new SlidingWindow<>(SLIDING_WINDOW_IN_SECONDS, TimeUnit.SECONDS);
+                        metricMap.put(indexShardKey, usageDeque);
+                    }
+                    usageDeque.next(new SlidingWindowData(this.clock.millis(), usage));
```

Comment:
We are gathering metrics at shard level here and at each iteration we are creating a new object. I wonder how much garbage we are creating here. We should try to reuse objects.
In the RCA periodic samplers, we should add a metric to get the gc count and duration metrics.

Reply:
Created an issue for this https://github.com/opendistro-for-elasticsearch/performance-analyzer-rca/issues/213